energymeters/energymeter.cpp

```diff
@@ -0,0 +1,206 @@
+{
+    return m_modbusRtuMaster->modbusUuid();
+}
+
+bool EnergyMeter::connected()
+{
+    return m_connected;
+}
+
+bool EnergyMeter::getVoltageL1()
+{
+    return getRegister(ModbusRegisterType::VoltageL1);
+}
+
+bool EnergyMeter::getVoltageL2()
+{
+    return getRegister(ModbusRegisterType::VoltageL2);
+}
+
+bool EnergyMeter::getVoltageL3()
+{
+    return getRegister(ModbusRegisterType::VoltageL3);
+}
+
+bool EnergyMeter::getCurrentL1()
+{
+    return getRegister(ModbusRegisterType::CurrentL1);
+}
+
+bool EnergyMeter::getCurrentL2()
+{
+    return getRegister(ModbusRegisterType::CurrentL2);
+}
+
+bool EnergyMeter::getCurrentL3()
+{
+    return getRegister(ModbusRegisterType::CurrentL3);
+}
+
+bool EnergyMeter::getFrequency()
+{
+    return getRegister(ModbusRegisterType::Frequency);
+}
+
+bool EnergyMeter::getPowerFactor()
+{
+    return getRegister(ModbusRegisterType::PowerFactor);
+}
+
+bool EnergyMeter::getActivePower()
+{
+    return getRegister(ModbusRegisterType::ActivePower);
+}
+
+bool EnergyMeter::getEnergyProduced()
+{
+    return getRegister(ModbusRegisterType::EnergyProduced);
+}
+
+bool EnergyMeter::getEnergyConsumed()
+{
+    return getRegister(ModbusRegisterType::EnergyConsumed);
+}
+
+bool EnergyMeter::getRegister(ModbusRegisterType type)
+{
+    if (!m_modbusRegisters.contains(type))
+        return false;
+
+    ModbusRegisterDescriptor descriptor = m_modbusRegisters.value(type);
+
+    ModbusRtuReply *reply = nullptr;
+    if (descriptor.functionCode() == 3){
+        reply = m_modbusRtuMaster->readHoldingRegister(m_slaveAddress, descriptor.address(), descriptor.length());
+    } else if (descriptor.functionCode() == 4){
+        reply = m_modbusRtuMaster->readInputRegister(m_slaveAddress, descriptor.address(), descriptor.length());
+    }
+    connect(reply, &ModbusRtuReply::finished, reply, &ModbusRtuReply::deleteLater);
+    connect(reply, &ModbusRtuReply::finished, this, [reply, type, descriptor, this] {
+        if (reply->error() != ModbusRtuReply::NoError) {
+            if (m_connected) {
+                m_connected = false;
+                emit connectedChanged(m_connected);
+            }
+            return;
+        } else {
+            if (!m_connected) {
+                m_connected = true;
+                emit connectedChanged(m_connected);
+            }
+        }
+        modbus_32_t value;
+        value.u = 0;
+        if (reply->result().length() == 1) {
+            value.u = static_cast<float>(reply->result().at(0));
+        } else if (reply->result().length() == 2) {
+            if (descriptor.dataType() == "float") {
+                value.u = static_cast<float>(static_cast<uint32_t>(reply->result().at(0)) << 16 | reply->result().at(1));
+            } else {
+                qCWarning(dcEnergyMeters()) << "Data type not supported" << descriptor.dataType();
+            }
+        } else {
+            return;
+        }
+
+        if (type == ModbusRegisterType::VoltageL1) {
```

Comment:
Handling this using a switch would probably be better, so you are always sure you handled all possible `type` values.

Reply:
Since you are so picky about performance ;) I stick with if else instead of switch, which is better in performance

Reply:
Performance wise probably, but the switch warns you if you forget to handle a possible value :-)